### PR TITLE
GROQ Mocks, and query optimizations

### PR DESCRIPTION
**Pull request by scottrippey**
# What
- Adds a working GROQ Mock, which efficiently executes GROQ queries
- Re-enables the "Products" page tests, which didn't work previously due to performance
- Optimizes several of our GROQ queries, by changing `[a && b]` into `[a][b]`

### CODE REVIEW

File: packages/nextjs/mocks/msw/groq/create-dataset.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  /**
+   * This list overrides what gets treated as an inline type.
+   */
+  inlineTypes?: string[];
+};
+
+/**
+ * Builds a flattened dataset from any type of deeply-nested data.
+ */
+export function createDataset(data: Array<AnyObject>, options: DatasetOptions = {}) {
```
Comment (scottrippey) [inline packages/nextjs/mocks/msw/groq/create-dataset.ts:26]: This is probably the highlight of the PR.
This takes any data, and maps it to Sanity's "dataset" structure. 

This is mostly just mapping objects into references, which is done automatically for data with `_id` attributes.

File: packages/nextjs/mocks/msw/groq/create-groq-handler.ts [hunk 2]
```diff
@@ -40,22 +52,22 @@ function parseParams(searchParams: object) {
...
+  const elapsed = Date.now() - start;
+  if (elapsed >= INEFFICIENT_QUERY_THRESHOLD) {
+    // Issue a warning!
+    console.warn(`
+      [groq-handler] WARNING: this query took ${elapsed} ms to mock execute.
+      This usually indicates an inefficient query, and you should consider improving it.
+      ${query.includes("&&") ? "Instead of using [a && b], consider using [a][b] instead!" : ""}
+      Inefficient query: \n${query}
+    `);
+  }
```
Comment (scottrippey) [inline packages/nextjs/mocks/msw/groq/create-groq-handler.ts:71]: Almost all of our queries took 15-20 seconds to execute.
By simply changing `[a && b]` to `[a][b]`, these queries went down to 1-3 seconds.

**Review by scottrippey — COMMENTED** [review]
Author's notes:

File: packages/nextjs/mocks/msw/groq/create-dataset.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  data.forEach((object) => builder.addData(object));
+  return builder.dataset;
+}
+
+class DatasetBuilder {
+  constructor(private options: DatasetOptions) {}
+
+  readonly dataset = [] as Dataset;
+
+  readonly references = new Map<object, DatasetItem>();
```
Comment (gksander) [inline packages/nextjs/mocks/msw/groq/create-dataset.ts:37]: Suppose there's any benefit to using a `WeakMap` here? If this acts something like a cache/temporary store, might be worthwhile (or maybe the GC affordances of a WeakMap don't really provide anything useful here, idk).

File: packages/nextjs/mocks/msw/groq/create-groq-handler.ts (hunk 2 above)
Comment (gksander) [reply packages/nextjs/mocks/msw/groq/create-groq-handler.ts:71]: Classic. I've found this to be the case even in production groq queries, so I now generally lean toward splitting up filters into something like `[a][b]` right from the get-go.

**Review by gksander — APPROVED** [review]
Neat!

File: packages/nextjs/mocks/msw/groq/create-dataset.ts (hunk 3 above)
Comment (scottrippey) [reply packages/nextjs/mocks/msw/groq/create-dataset.ts:37]: This class is a "throwaway" class; the `createDataset` method initializes it, uses it, and throws it away. So the references are discarded as soon as the results are calculated.  
I _think_ that `Map` is minutely faster than `WeakMap` so it's probably the best choice for this purpose.

**Pull request merged**